lib/eval.ml

```diff
@@ -699,210 +605,42 @@ module MakeInterpreter (T : Target) = struct
-    | _ -> failwith "not an lvalue"
-
-  and typ_of_lvalue (ctrl: ctrl) (env: env) (st: state) (lv : lvalue) : Type.t =
-    match lv with
-    | LName s                            -> EvalEnv.find_typ s env
-    | LMember{expr=lv';name=s}           -> typ_of_lmember ctrl env st lv' s
-    | LBitAccess{expr=lv';msb=e1;lsb=e2} -> typ_of_lbit ctrl env st lv' e1 e2
-    | LArrayAccess{expr=lv';idx=e}       -> typ_of_larray ctrl env st lv' e
-
-  and typ_of_lmember ctrl env st (lv : lvalue) (s : string) : Type.t =
-    let t = typ_of_lvalue ctrl env st lv in
-    match snd t with
-    | HeaderStack{header;_} -> typ_of_stack_lmem env ctrl st s header
-    | TypeName name ->
-      begin match snd (decl_of_typ env t) with
-        | Header{fields=fs;_}      -> typ_of_struct_lmem s fs
-        | HeaderUnion{fields=fs;_} -> typ_of_struct_lmem s fs
-        | Struct{fields=fs;_}      -> typ_of_struct_lmem s fs
-        | _ -> failwith "lvalue type name member access not defined" end
-    | _ -> failwith "type of lvalue member unimplemented"
-
-  and typ_of_struct_lmem (s : string) (fields : Declaration.field list) : Type.t =
-    let fs = List.map fields ~f:(fun a -> (snd (snd a).name, a)) in
-    let f = List.Assoc.find_exn fs ~equal:String.equal s in
-    (snd f).typ
-
-  and typ_of_stack_lmem env ctrl st (s : string) (t : Type.t) : Type.t =
-    let () =
-      match s with
-      | "next" -> ()
-      | _ -> failwith "stack member not a lvalue" in
-    t
-
-  and typ_of_lbit (ctrl : ctrl) (env : env) (st : state) (lv : lvalue) (e1 : Expression.t)
-      (e2 : Expression.t) : Type.t =
-    let (_,_,_,v1) = eval_expr ctrl env st SContinue e1 in
-    let (_,_,_,v2) = eval_expr ctrl env st SContinue e2 in
-    let n1 = bigint_of_val v1 in
-    let n2 = bigint_of_val v2 in
-    let n0 = Bigint.(n1 - n2 + one) in
-    (Info.dummy, Type.BitType(Info.dummy, Expression.Int(Info.dummy,{value=n0;width_signed=None})))
-
-  and typ_of_larray (ctrl : ctrl) (env : env) (st : state) (lv : lvalue) (e : Expression.t) : Type.t =
-    let t = typ_of_lvalue ctrl env st lv in
-    match snd t with
-    | HeaderStack{header;_} -> header
-    | _ -> failwith "array access on non-header stack"
+  and lvalue_of_expr (ctrl : ctrl) (env : env) (st : state) (signal : signal)
+      (expr : Expression.t) : env * state * signal * lvalue =
+    match signal with
+    | SContinue -> begin match (snd expr).expr with
+      | Name name -> env, st, SContinue, LName {name; typ = (snd expr).typ}
+      | ExpressionMember{expr=e; name=(_,n)} -> lvalue_of_expr_mem ctrl env st (snd expr).typ e n
+      | BitStringAccess{bits;lo;hi} -> lvalue_of_expr_bsa ctrl env st (snd expr).typ bits lo hi
+      | ArrayAccess{array=a;index} -> lvalue_of_expr_ara ctrl env st (snd expr).typ a index
+      | _ -> failwith "not an lvalue" end
+    | SReject _ | SExit | SReturn _ -> env, st, signal, LName {name = BareName (Info.dummy, ""); typ = Void}
```

Comment:
Can this function return an optional expression instead of a variable named `""`

Reply:
done